data_model/src/events/data/filters.rs

```diff
@@ -0,0 +1,711 @@
+            matches!(
+                (self, event),
+                (Self::ByCreated, TriggerEvent::Created(_))
+                    | (Self::ByDeleted, TriggerEvent::Deleted(_))
+                    | (Self::ByExtended, TriggerEvent::Extended(_))
+                    | (Self::ByShortened, TriggerEvent::Shortened(_))
+            )
+        }
+    }
+}
+
+/// Filter for all events
+pub type EventFilter = FilterOpt<EntityFilter>;
+
+/// Trait for filters
+pub trait Filter {
+    /// Type of event that can be filtered
+    type EventType;
+
+    /// Check if `item` matches filter
+    ///
+    /// Returns `true`, if `item` matches filter and `false` if not
+    fn matches(&self, item: &Self::EventType) -> bool;
+}
+
+#[derive(
+    Clone,
+    PartialEq,
+    PartialOrd,
+    Ord,
+    Eq,
+    Debug,
+    Decode,
+    Encode,
+    Deserialize,
+    Serialize,
+    IntoSchema,
+    Hash,
+)]
+/// Optional filter. May pass all items or may filter them by `F`
+///
+/// It's better than `Optional<F>` because `Optional` already has its own `filter` method and it
+/// would be ugly to use fully qualified syntax to call `Filter::filter()` method on it.
+/// Also `FilterOpt` variant names look better for filter needs
+pub enum FilterOpt<F: Filter> {
+    /// Accept all items that will be passed to `filter()` method
+    AcceptAll,
+    /// Use filter `F` to choose acceptable items passed to `filter()` method
+    BySome(F),
+}
+
+impl<F: Filter> Filter for FilterOpt<F> {
+    type EventType = F::EventType;
+
+    fn matches(&self, item: &Self::EventType) -> bool {
+        match self {
+            Self::AcceptAll => true,
+            Self::BySome(filter) => filter.matches(item),
+        }
+    }
+}
+
+#[derive(
+    Clone,
+    PartialEq,
+    PartialOrd,
+    Ord,
+    Eq,
+    Debug,
+    Decode,
+    Encode,
+    Deserialize,
+    Serialize,
+    FromVariant,
+    IntoSchema,
+    Hash,
+)]
+#[allow(clippy::enum_variant_names)]
+/// Filters event by entity
+pub enum EntityFilter {
+    /// Filter by Domain entity. `AcceptAll` value will accept all `Domain` events
+    ByDomain(FilterOpt<DomainFilter>),
+    /// Filter by Peer entity. `AcceptAll` value will accept all `Peer` events
+    ByPeer(FilterOpt<PeerFilter>),
+    /// Filter by Role entity. `AcceptAll` value will accept all `Role` events
+    #[cfg(feature = "roles")]
+    ByRole(FilterOpt<RoleFilter>),
+    /// Filter by Account entity. `AcceptAll` value will accept all `Account` events
+    ByAccount(FilterOpt<AccountFilter>),
+    /// Filter by AssetDefinition entity. `AcceptAll` value will accept all `AssetDefinition` events
+    ByAssetDefinition(FilterOpt<AssetDefinitionFilter>),
+    /// Filter by Asset entity. `AcceptAll` value will accept all `Asset` events
+    ByAsset(FilterOpt<AssetFilter>),
+    /// Filter by Trigger entity. `AcceptAll` value will accept all `Trigger` events
+    ByTrigger(FilterOpt<TriggerFilter>),
+}
```

Comment:
seeing `FilterOpt<DomainFilter>` and alike inside of every variant makes me think that it can be extracted to `FilterOpt<EntityFilter>

Reply:
Well, yeah, a lot of `FilterOpt`, but this gives a lot of flexibility. For example:

```rust
let filter1 = AcceptAll; // Accepts all events

let filter2 = BySome(EntityFilter::ByDomain(AcceptAll)); // Accepts all domain events

// Accepts all events of account with id == `account_id`
let filter3 = BySome(EntityFilter::ByAccount(BySome(AccountFilter::new(
    BySome(IdFilter(account_id)),
    AcceptAll,
))));
```

Reply:
We can do some optimizations and refactoring later, if you have a suggestion